parasol_cpu/src/proc/assembly.rs

```diff
@@ -43,9 +44,81 @@ macro_rules! define_op {
                     ),*
                 }
 
+                impl TryFrom<u64> for $inst_name {
+                    type Error = Error;
+
+                    fn try_from(value: u64) -> $crate::Result<Self> {
+                        #[allow(unused_variables)]
+                        Ok(match [<$inst_name OpCode>]::try_from((value & 0xFF) as u8)? {
+                            $(
+                                [<$inst_name OpCode>]::$op_name => {
+                                    #[allow(unused)]
+                                    let value = value >> 8;
+
+                                    $(
+                                        let $dst_name = RegisterName::new(value as usize & 0x1F);
+                                        #[allow(unused)]
+                                        let value = value >> 5;
+                                    )*
+
+                                    $(
+                                        let $src_name = RegisterName::new(value as usize & 0x1F);
+                                        #[allow(unused)]
+                                        let value = value >> 5;
+                                    )*
+
+                                    $(
+                                        let mask = (0x1u64 << $meta_width) - 1;
+                                        let $meta_name = (value & mask) as $meta_type;
+                                        let value = value >> $meta_width;
+                                    )*
+
+                                    $inst_name :: $op_name(
+                                        $($dst_name,)*
+                                        $($src_name,)*
+                                        $($meta_name,)*
+                                    )
+                                }
+                            )*
+                        })
+                    }
+                }
+
+                impl From<$inst_name> for u64 {
+                    fn from(x: $inst_name) -> Self {
+                        let mut encoded = x.op_code() as u64;
+                        let mut shift = 8;
+
+                        #[allow(unused_assignments)]
+                        match x {
+                            $(
+                                $inst_name::$op_name (
+                                    $($dst_name,)*
+                                    $($src_name,)*
+                                    $($meta_name,)*
+                                ) => {
+                                    $(
+                                        encoded |= ($dst_name.name as u64) << shift;
+                                        shift += 5;
```

Comment:
Parasol today has 64 registers in the encoded instruction, not 32. So 6, not 5. Ideally, we pass the count and derive this value.